src/Orleans.Core/Runtime/CallbackData.cs

```diff
@@ -183,37 +119,31 @@ private void OnFail(Message msg, Message error, string resendLogMessageFormat, b
                 if (alreadyFired)
                     return;
 
-                if (messagingOptions.ResendOnTimeout && resendFunc(msg))
+                if (this.shared.MessagingOptions.ResendOnTimeout && this.shared.ShouldResend(msg))
                 {
-                    if (logger.IsEnabled(LogLevel.Debug)) logger.Debug(resendLogMessageFormat, msg.ResendCount, msg);
+                    if (this.shared.Logger.IsEnabled(LogLevel.Debug)) this.shared.Logger.Debug(resendLogMessageFormat, msg.ResendCount, msg);
                     return;
                 }
 
                 alreadyFired = true;
-                DisposeTimer();
                 if (StatisticsCollector.CollectApplicationRequestsStats)
                 {
-                    timeSinceIssued.Stop();
+                    durationTimestamp = -(Stopwatch.GetTimestamp() - durationTimestamp);
```

Comment:
Nit: Repetitive calculation (duplication of line 102) could be extracted into separate method.

Reply:
Yeah, that's true. I was wary of doing that because of how I'm abusing that single field to mean two things depending on the sign